ts/session/sending/MessageQueue.ts

```diff
@@ -1,54 +1,186 @@
-    throw new Error('Method not implemented.');
+
+  public async sendMessageToDevices(
+    devices: Array<PubKey>,
+    message: ContentMessage
+  ) {
+    let currentDevices = [...devices];
+
+    // Sync to our devices if syncable
+    if (SyncMessageUtils.canSync(message)) {
+      const currentDevice = await UserUtils.getCurrentDevicePubKey();
+
+      if (currentDevice) {
+        const otherDevices = await getPairedDevicesFor(currentDevice);
+
+        const ourDevices = [currentDevice, ...otherDevices].map(
+          device => new PubKey(device)
+        );
+        await this.sendSyncMessage(message, ourDevices);
+
+        // Remove our devices from currentDevices
+        const ourDeviceContacts = ourDevices.map(device =>
+          ConversationController.get(device.key)
+        );
+        currentDevices = _.xor(currentDevices, ourDeviceContacts);
+      }
+    }
+
+    const promises = currentDevices.map(async device => {
+      await this.process(device, message);
+    });
+
+    return Promise.all(promises);
   }
-  public sendSyncMessage(message: ContentMessage) {
-    throw new Error('Method not implemented.');
+
+  public async sendToGroup(
+    message: OpenGroupMessage | ContentMessage
+  ): Promise<boolean> {
+    if (
+      !(message instanceof OpenGroupMessage) &&
+      !(message instanceof ClosedGroupMessage)
+    ) {
+      return false;
+    }
+
+    // Closed groups
+    if (message instanceof ClosedGroupMessage) {
+      // Get devices in closed group
+      const conversation = ConversationController.get(message.groupId);
+      const recipientsModels = conversation.contactCollection.models;
+      const recipients: Array<PubKey> = recipientsModels.map(
+        (recipient: any) => new PubKey(recipient.id)
+      );
+
+      await this.sendMessageToDevices(recipients, message);
+
+      return true;
+    }
+
+    // Open groups
+    if (message instanceof OpenGroupMessage) {
+      // No queue needed for Open Groups; send directly
+
+      try {
+        await MessageSender.sendToOpenGroup(message);
+        this.events.emit('success', message);
+      } catch (e) {
+        this.events.emit('fail', message, e);
+      }
+
+      return true;
+    }
+
+    return false;
   }
 
-  public processPending(device: string) {
-    // TODO: implement
+  public async sendSyncMessage(message: ContentMessage, sendTo: Array<PubKey>) {
+    // Sync with our devices
+    const promises = sendTo.map(async device => {
+      const syncMessage = SyncMessageUtils.from(message);
+
+      return this.process(device, syncMessage);
+    });
+
+    return Promise.all(promises);
   }
 
-  private processAllPending() {
-    // TODO: Get all devices which are pending here
+  public async processPending(device: PubKey) {
+    const messages = this.pendingMessageCache.getForDevice(device);
+
+    // const isMediumGroup = messages.some(m => m instance of MediumGroupMessage)
+    const isMediumGroup = false;
```

Comment:
@msgmaxim how do you tell a medium group apart from a regular user? and is there a quick util we can use to determine if a user is a medium group without relying on window?

Reply:
this quick work around just assumes a MediumGroupMessage means we're sending to a medium group, this might work but if there's a better way rather than importing `ConversationController` from window then it would be better.